File: Packages/EditorKit/Tests/EditorKitTests/ArrowGeometryTests.swift
import TestKit
import CoreGraphics
@testable import EditorKit

let arrowGeometryTests: [TestCase] = [
    TestCase("horizontalArrowheadWings") { t in
        // Arrow pointing right: start (0,0) → end (100,0), head length 10, half-angle 30°.
        let (left, right) = ArrowGeometry.headWings(
            start: CGPoint(x: 0, y: 0), end: CGPoint(x: 100, y: 0),
            length: 10, halfAngleDegrees: 30)
        // Wings sit behind the tip (x < 100) and symmetric about y=0.
        t.isTrue(left.x < 100)
        t.isTrue(right.x < 100)
        t.approxEqual(Double(left.y), Double(-right.y), tol: 0.001)
        t.approxEqual(Double(abs(left.y)), Double(10 * sin(30 * CGFloat.pi / 180)), tol: 0.01)
    },
    TestCase("shaftEndStopsAtArrowheadBase") { t in
        // Horizontal arrow (0,0)→(100,0), head length 30, half-angle 28°.
        // The shaft must end one head-depth (length·cos28°) short of the tip.
        let e = ArrowGeometry.shaftEnd(start: CGPoint(x: 0, y: 0),
                                       end: CGPoint(x: 100, y: 0),
                                       headLength: 30, halfAngleDegrees: 28)
        t.approxEqual(Double(e.x), Double(100 - 30 * cos(28 * CGFloat.pi / 180)), tol: 0.01)
        t.approxEqual(Double(e.y), 0, tol: 0.001)
        t.isTrue(e.x < 100) // never reaches the tip
    },
    TestCase("shaftEndClampsToStartForShortArrow") { t in
        // Arrow shorter than the head: the shaft collapses to the start (no
        // negative-length shaft pointing backwards).
        let e = ArrowGeometry.shaftEnd(start: CGPoint(x: 0, y: 0),
                                       end: CGPoint(x: 10, y: 0),
                                       headLength: 30, halfAngleDegrees: 28)
        t.approxEqual(Double(e.x), 0, tol: 0.001)
        t.approxEqual(Double(e.y), 0, tol: 0.001)
    },
]

File: windows/tests/BetterScreenshot.Tests/ArrowGeometryTests.cs
using BetterScreenshot.Core;
using BetterScreenshot.Editor;
using Xunit;

namespace BetterScreenshot.Tests;

public class ArrowGeometryTests
{
    [Fact]
    public void HorizontalArrowheadWings()
    {
        var (left, right) = ArrowGeometry.HeadWings(new PxPoint(0, 0), new PxPoint(100, 0), length: 20, halfAngleDegrees: 28);
        // Symmetric across the shaft (x-axis), same X, and behind the tip.
        Assert.Equal(left.X, right.X, 6);
        Assert.Equal(left.Y, -right.Y, 6);
        Assert.True(left.X < 100);
        // Each wing is `length` from the tip.
        Assert.Equal(20, Dist(new PxPoint(100, 0), left), 6);
        Assert.Equal(20, Dist(new PxPoint(100, 0), right), 6);
    }

    [Fact]
    public void ShaftEndStopsAtArrowheadBase()
    {
        var s = ArrowGeometry.ShaftEnd(new PxPoint(0, 0), new PxPoint(100, 0), headLength: 20, halfAngleDegrees: 28);
        double expected = 100 - 20 * Math.Cos(28 * Math.PI / 180.0); // 82.34
        Assert.Equal(expected, s.X, 6);
        Assert.Equal(0, s.Y, 6);
    }

    [Fact]
    public void ShaftEndClampsToStartForShortArrow()
    {
        var start = new PxPoint(0, 0);
        var s = ArrowGeometry.ShaftEnd(start, new PxPoint(10, 0), headLength: 20, halfAngleDegrees: 28);
        Assert.Equal(start, s);
    }

    private static double Dist(PxPoint a, PxPoint b) => Math.Sqrt(Math.Pow(a.X - b.X, 2) + Math.Pow(a.Y - b.Y, 2));
}
